File: backend/models/bond.py
from datetime import date
from typing import Optional
from pydantic import BaseModel, field_validator


class StepUp(BaseModel):
    start_date: date
    end_date: date
    coupon_delta: float        # annual delta as decimal, e.g. 0.005 = +50 bps; negative = step-down
    probability: float = 0.5  # 0.0 – 1.0

    @field_validator("probability")
    @classmethod
    def prob_range(cls, v: float) -> float:
        if not 0.0 <= v <= 1.0:
            raise ValueError("probability must be between 0 and 1")
        return v


class CallOption(BaseModel):
    call_date: date
    call_price: float = 100.0   # redemption price, usually par


class YieldCurvePoint(BaseModel):
    tenor: str   # "0d" | "1m" | "3m" | "6m" | "1y" | "2y" | "3y" | "5y" | "7y" | "10y" | "15y" | "20y"
    rate: float  # annual rate as decimal


class BondPriceRequest(BaseModel):
    settlement_date: date
    maturity_date: date
    face_value: float = 100.0
    coupon_rate: float
    coupon_frequency: int = 2
    day_count: str = "ACT/ACT"
    yield_rate: Optional[float] = None   # basic mode; omit in advanced mode
    step_ups: list[StepUp] = []
    call_option: Optional[CallOption] = None
    yield_curve: Optional[list[YieldCurvePoint]] = None
    z_spread: Optional[float] = 0.0     # advanced mode: constant spread added to every curve rate

    @field_validator("coupon_frequency")
    @classmethod
    def valid_frequency(cls, v: int) -> int:
        if v not in (1, 2, 4, 12):
            raise ValueError("coupon_frequency must be 1, 2, 4, or 12")
        return v


class ScenarioResult(BaseModel):
    label: str
    coupon_delta: float
    probability: float
    price: float
    clean_price: float
    pv_of_stepup: float
    certain_clean_price: Optional[float] = None  # clean price if this scenario occurs with certainty
    spread_value_bps: Optional[float] = None  # bps the step-up is worth at the input z-spread


class CashFlowEntry(BaseModel):
    date: date
    base_coupon: float
    expected_coupon: float
    principal: float
    discount_factor: float
    base_coupon_pv: float
    expected_coupon_pv: float
    principal_pv: float
    curve_rate: Optional[float] = None


class BondPriceResponse(BaseModel):
    base_price: float
    expected_price: float
    base_clean_price: float
    expected_clean_price: float
    accrued_interest: float
    scenario_results: list[ScenarioResult]
    cashflow_schedule: list[CashFlowEntry]
    price_to_call: Optional[float] = None
    clean_price_to_call: Optional[float] = None
    step_up_spread_bps: Optional[float] = None  # bps all step-ups combined are worth at input z-spread

File: frontend/src/types/bond.ts
export type DayCount = 'ACT/ACT' | 'ACT/360' | 'ACT/365' | '30/360'
export type Frequency = 1 | 2 | 4 | 12
export type StepType = 'coupon_delta' | 'principal_pct'

export interface StepUp {
  start_date: string   // ISO 8601 YYYY-MM-DD
  end_date: string
  coupon_delta: number // annual decimal; meaning depends on step_type
  probability: number  // 0–1
  step_type: StepType  // 'coupon_delta' = bps change to coupon rate; 'principal_pct' = % of outstanding principal
}

export interface CallOption {
  call_date: string
  call_price: number
}

export interface YieldCurvePoint {
  tenor: string  // "0d" | "1m" | "3m" | "6m" | "1y" | "2y" | "3y" | "5y" | "7y" | "10y" | "15y" | "20y"
  rate: number   // decimal
}

export interface BondPriceRequest {
  settlement_date: string
  maturity_date: string
  face_value: number
  coupon_rate: number       // decimal
  coupon_frequency: Frequency
  day_count: DayCount
  yield_rate?: number       // decimal; omitted in advanced (curve) mode
  step_ups: StepUp[]
  call_option?: CallOption | null
  yield_curve?: YieldCurvePoint[]
  z_spread?: number         // decimal; advanced mode only (e.g. 0.01 = 100 bps)
}

// ---- Response types ----

export interface ScenarioResult {
  label: string
  coupon_delta: number
  probability: number
  price: number
  clean_price: number
  pv_of_stepup: number
  certain_clean_price?: number
  spread_value_bps?: number
  yield_value_bps?: number
}

export interface CashFlowEntry {
  date: string
  base_coupon: number
  expected_coupon: number
  principal: number
  outstanding_principal: number
  discount_factor: number
  base_coupon_pv: number
  expected_coupon_pv: number
  principal_pv: number
  curve_rate?: number
}

export interface BondPriceResponse {
  base_price: number
  expected_price: number
  base_clean_price: number
  expected_clean_price: number
  accrued_interest: number
  scenario_results: ScenarioResult[]
  cashflow_schedule: CashFlowEntry[]
  has_principal_pct_steps?: boolean
  price_to_call: number | null
  clean_price_to_call: number | null
  step_up_spread_bps?: number
  step_up_yield_bps?: number
}
